semantic-core/src/Data/Scope.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE DeriveTraversable, LambdaCase, RankNTypes, QuantifiedConstraints, StandaloneDeriving #-}
+module Data.Scope
+( Incr(..)
+, incr
+, Scope(..)
+, fromScope
+, toScope
+, bind1
+, bind
+, bindEither
+, instantiate1
+, instantiate
+, instantiateEither
+) where
+
+import Control.Applicative (liftA2)
+import Control.Effect.Carrier
+import Control.Monad ((>=>), guard)
+import Control.Monad.Module
+import Control.Monad.Trans.Class
+import Data.Function (on)
+
+data Incr a b
+  = Z a
+  | S b
+  deriving (Eq, Foldable, Functor, Ord, Show, Traversable)
+
+instance Applicative (Incr a) where
+  pure = S
+  Z a <*> _ = Z a
+  S f <*> a = f <$> a
+
+instance Monad (Incr a) where
+  Z a >>= _ = Z a
+  S a >>= f = f a
+
+match :: Applicative f => (b -> Either a c) -> b -> Incr a (f c)
+match f x = either Z (S . pure) (f x)
+
+matchMaybe :: (b -> Maybe a) -> (b -> Either a b)
+matchMaybe f a = maybe (Right a) Left (f a)
+
+incr :: (a -> c) -> (b -> c) -> Incr a b -> c
+incr z s = \case { Z a -> z a ; S b -> s b }
+
+
+newtype Scope a f b = Scope { unScope :: f (Incr a (f b)) }
+  deriving (Foldable, Functor, Traversable)
+
+instance HFunctor (Scope a) where
+  hmap f = Scope . f . fmap (fmap f) . unScope
+
+instance (Eq   a, Eq   b, forall a . Eq   a => Eq   (f a), Monad f) => Eq   (Scope a f b) where
+  (==) = (==) `on` fromScope
```

Comment:
Unlike “normal” de Bruijn-indexed representations (`f (Incr a)`), Paterson’s extended representation (`f (Incr (f a))`) can represent alpha-equivalent terms in multiple ways. We therefore define specialized `Eq` and `Ord` instances so that `==` and `compare` respect alpha-equivalence.

`fromScope` can be seen as canonicalizing the extended representation to the basic one.